Update Write flows page (#14610)
PrefectHQ/prefect: docs/3.0rc/develop/write-flows.mdx

description: Learn the basics of defining and running flows.
---

Flows are the most central Prefect objects. 
A flow is a container for workflow logic as code. 

Flows are defined as Python functions. 
They can take inputs, perform work, and return an output.

> Turn any function into a Prefect flow by adding the `@flow` decorator to it:

Reviewer: ```suggestion
Make any function a Prefect flow by adding the `@flow` decorator to it:
```

    return
```

When a function becomes a flow, its behavior changes. 
Flows have the following capabilities:

- All runs of the flow have persistent [state](/3.0rc/develop/manage-states/). Transitions between states are recorded, 
allowing you to observe and act on flow execution.

> - Input arguments can be type validated as workflow parameters.

Reviewer: ```suggestion
- Input arguments can be type validated as workflow [parameters](/#specify-flow-parameters).
```

If you don't provide a name, Prefect uses the flow function name.

## Run flows

A _flow run_ is a single execution of a flow.

> Create a flow run by calling a flow by its function name, just as you would a normal Python function in a script.

Reviewer: ```suggestion
Create a flow run by calling a flow by its function name, just as you would a normal Python function.
```

- Starting a flow run for the deployment through a schedule, the Prefect UI, or the Prefect API

However you run your flow, Prefect monitors the flow run, capturing its state for observability.
You can log a [variety of metadata](/3.0rc/develop/logging) about flow runs for monitoring, troubleshooting, and auditing purposes.

> The script below uses the HTTPX client library to fetch statistics about the [main Prefect repository](https://github.com/PrefectHQ/prefect).

Reviewer: ```suggestion
The example below uses the HTTPX client library to fetch statistics about the [main Prefect repository](https://github.com/PrefectHQ/prefect).
```

```python
from prefect import flow

def gen():
    yield from [1, 2, 3]
    print('Generator consumed!')

@flow
def f():
    yield gen

generator = next(f())
list(generator) # prints 'Generator consumed!'
```
</Warning>

## Flow runs

> A _flow run_ represents a single execution of a flow.

Reviewer: ```suggestion
A _flow run_ is a single execution of a flow.
```

> In adddition to create a flow run by calling the flow manually, you can create a flow run by using an external scheduler such as `cron` to invoke a flow function. 

Reviewer: ```suggestion
You can create a flow run by calling the flow function manually, or even by using an external scheduler such as `cron` to invoke a flow function. 
```

> More commonly, most users run flows by create a [deployment](/3.0rc/deploy/) on Prefect Cloud or Prefect server and then scheduling a flow run for the deployment through a schedule, the Prefect UI, or the Prefect API.

Reviewer: ```suggestion
Most users run flows by creating a [deployment](/3.0rc/deploy/) on Prefect Cloud or Prefect server and then scheduling a flow run for the deployment through a schedule, the Prefect UI, or the Prefect API.
```

- If the flow does not return a value (or returns `None`), its state is determined by the states of all of the tasks and nested flows within it.
  - If _any_ task run or nested flow run fails, then the final flow run state is marked as `FAILED`.
  - If _any_ task run is cancelled, then the final flow run state is marked as `CANCELLED`.
- If a flow returns a manually created state, it is used as the state of the final flow run. This allows for manual determination of final state.
- If the flow run returns _any other object_, then it is marked as completed.

### Return a future

If a flow returns one or more futures, the final state is determined based on the underlying states.

```python
from prefect import flow, task

@task
def always_fails_task():
    raise ValueError("I fail successfully")

@task
def always_succeeds_task():
    print("I'm fail safe!")
    return "success"

@flow
def always_succeeds_flow():
    x = always_fails_task.submit().result(raise_on_failure=False)
    y = always_succeeds_task.submit(wait_for=[x])
    return y

if __name__ == "__main__":
    always_succeeds_flow()
```

> Running this flow results in a **Completed** final state because the flow returns the future of the task that succeeds:

Reviewer: ```suggestion
This flow run finishes in a **Completed** final state because the flow returns the future of the task that succeeds:
```

I'm fail safe!
07:29:35.086 | INFO    | Task run 'always_succeeds_task-0' - Finished in state Completed()
07:29:35.225 | INFO    | Flow run 'hidden-butterfly' - Finished in state Completed('I am happy with this result')
```

If a flow run returns any other object, then it is recorded as `COMPLETED`

## Retries

Unexpected errors may occur in workflows. 
For example the GitHub API may be temporarily unavailable or rate limited.

Prefect can automatically retry flow runs on failure. 

To enable retries, pass an integer to the flow's `retries` parameter.
If the flow run fails, Prefect will retry it up to `retries` times.

If the flow run fails on the final retry, Prefect records the final flow run state as _failed_.

Optionally, pass an integer to `retry_delay_seconds` to specify how many seconds to wait between each retry attempt. 

Check out [Transactions](/3.0rc/develop/transactions/) to make your flows even more resilient and rollback actions when desired.

## See also

- Store and reuse non-sensitive bits of data, such as configuration information, by using [variables](/3.0rc/develop/variables).

> - Organize a flow with [tasks](/3.0rc/develop/write-tasks/) to break down the workflow's complexity and make it more performant and observable.

Reviewer: ```suggestion
- Make your flow more manageable, performant, and observable by breaking it down into discrete units of orchestrated work with [tasks](/3.0rc/develop/write-tasks/).
```

Author: Thank you. Removed "down".
